# Generate chromosome genlight objects with adegenet

# Please note: There are many ways to generate `genlight` objects. We provide 
#  this code as a potential starting point for anyone interested in generating 
#  similar data objects. Be aware, though, that the script is not optimized
#  for speed or memory efficiency, and, while it was written to be as general
#  as possible, it may require some alterations to make it work on different 
#  computational setups.

# Script can be run using: `Rscript ARG1 ARG2 ...` with ARG1 being the 
#  arguments described below

# The script uses some custom functions from `custom_sg_analysis_functions.r`,
#  the path to this file must be set in `custom_func_file` in the script

# For our analysis, the total VCF was divided by chromosome, and each
#  chromosome was divided into subfiles to accommodate memory limitations in R.
#  We split the VCFs into sub-files containing 100k SNPs.

# This script was written to process multiple divided VCF files from a single
#  chromosome, with names that are in SNP order when using `ls` in Linux,
#  though it could be used to process any set of divided VCF files

# The script uses a separate file containing the column header information for 
#  the VCF, including the sample names, to make processing the VCFs easier. 
#  This file can be generated by simply coping that line that line from the 
#  VCF to a new file. ex: `grep #POS FULL_VCF_FILE.vcf > vcf_header.txt`

# This script requires separate files with the sample names for 4X and 8X 
#  samples - these are processed differently: 4X receive disomic genotypes,
#  8X receive tetrasomic genotypes

############

# args[1] = data_dir = Directory containing the divided VCFs

# args[2] = vcf_search_string = a string that can be used find the divided 
#                                 VCFs for the chromosome, using the `ls` 
# 				  command in Linux
#             			  ex: 'Chr01K.vcf_*'

# args[3] = vcf_header_file = Full path and file name of the file with the 
#				column header info for the VCF; see notes 
#				above for more information about this file

# args[4] = tet_lib_file = Full path and file name of file with 4X sample names
#			    as written in the vcf header information

# args[5] = oct_lib_file = Full path and file name of file with 8X sample names
#                           as written in the vcf header information

# args[6] = out_name_short = the name of the final file; 
# 			       ex: Chr01K.genlight.rds

# args[7] = maf_cut_in = the minor allele frequency cutoff; ex: 0.002


###############

args = commandArgs(trailingOnly = TRUE)

### LOAD PACKAGES ###

library(adegenet)
library(parallel)

# Load custom functions
custom_func_file <- '/PATH_TO_FILE/custom_sg_analysis_functions.r'
source(custom_func_file)

### IMPORT DATA ###

data_dir <- args[1]
data_dir <- add_slash(data_dir)

vcf_search_string <- args[2]
vcf_search_command <- paste('ls ', data_dir, vcf_search_string, sep = '')

vcf_files <- system(vcf_search_command, inter = T)

# load first VCF file to start the total VCF
vcf_1 <- read.table(vcf_files[1], header = F, stringsAsFactors = F, 
    sep = '\t')

vcf_header_file <- args[3]
vcf_header <- gsub('#', '', read.table(vcf_header_file, stringsAsFactors = F,
  sep = '\t', header = F, comment.char = '@'))

colnames(vcf_1) <- vcf_header

tet_lib_file <- args[4] 
tet_libs_0 <- as.vector(read.table(tet_lib_file, header = F,
  stringsAsFactors = F)[,1])
tet_libs_1 <- intersect(tet_libs_0, vcf_header)

oct_lib_file <- args[5] 
oct_libs_0 <- as.vector(read.table(oct_lib_file, header = F,
  stringsAsFactors = F)[,1])
oct_libs_1 <- intersect(oct_libs_0, vcf_header)

### SET OUTPUTS ###
out_name_short <- args[6] 
out_name_full <- paste(data_dir, out_name_short, sep = '')

### SET VARIABLES ###
maf_cut_in <- args[7]
maf_cut_in <- as.numeric(maf_cut_in)
###########################

preobj_1 <- gen_gl_preobj(vcf = vcf_1, oct_libs = oct_libs_1, 
  tet_libs = tet_libs_1, maf_cut = maf_cut_in)

gl_tot <- gen_gl_object(preobj_list = preobj_1)

if(length(vcf_files) > 1){
  for(i in c(2:length(vcf_files))){
    vcf_tmp <- read.table(vcf_files[i], header = T, stringsAsFactors = F,
      sep = '\t')
    colnames(vcf_tmp) <- vcf_header
    preobj_tmp <- gen_gl_preobj(vcf = vcf_tmp, oct_libs = oct_libs_1, 
      tet_libs = tet_libs_1, maf_cut = maf_cut_in)
    gl_tmp <- gen_gl_object(preobj_list = preobj_tmp)
    gl_tot <- cbind(gl_tot, gl_tmp)
    print(i)
  }
}

saveRDS(gl_tot, out_name_full)

quit(save = 'no')

